### PR TITLE
Split SpanContext into interface / impl

### PR DESCRIPTION
**Pull request by anuraaga**
As enabled by https://github.com/open-telemetry/opentelemetry-specification/pull/969

### CODE REVIEW

**Comment by Oberon00** [comment]
This seems to make the code more complicated and arguably does not implement the "SHOULD" in "The API MUST implement methods to create a SpanContext. These methods SHOULD be the only way to create a SpanContext." since you can now create a SpanContext by deriving from the interface and instantiate the derived class.
So can you please state the reasons for these changes here?

File: api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java
Comment (Oberon00) [inline api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java:14]: Since the interface is also annotated `@Immutable`  and according to the spec any SpanContext must be Immutable, this is maybe not the best name to distinguish it from possible other implementations which would also be Immutable. Maybe DefaultSpanContext?

**Comment by anuraaga** [comment]
@Oberon00 Weren't all the reasons in the long discussion in https://github.com/open-telemetry/opentelemetry-specification/pull/969 and https://github.com/open-telemetry/opentelemetry-specification/issues/970 ? :-) The short answer is for agent users, it will always be the agent's version of the API, not the user's version of the API, that instantiates a `SpanContext`. I thought we agreed this is ok in that issue.

Also one point of note is that we don't lose any protection with this PR since it's already an abstract class. Of course, we could stop using AutoValue to prevent that, just an observation. With #1937 I am thinking of a general question on whether we should just avoid classes at all in the API - it's an API after all. If a custom SDK (which our auto instrumentation agent effectively is) finds a reason to override the implementation and satisfies a use case with it, why block it?

File: api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java
Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java:14]: Yeah I had the same impression - we have `Immutable*` as a pattern in many AutoValue implementations of interfaces currently. Renaming them all to `Default` seems fine to me, or even `*Impl` so the implementation orders next to the interface in the filetree.

**Comment by Oberon00** [comment]
I suspected as much 😃 But I thought it was better to document it on the PR that does the change.

So the interface makes it even easier for the auto-instrumentation than the previous abstract class?

**Comment by anuraaga** [comment]
@Oberon00 Yeah - one class can implement both the agent API interface and application API interface, but with abstract classes we wouldn't be able to do that.

File: api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java:14]: No strong opinion. I'd keep this name for this PR, and we can always rename later, since this is an internal detail.

**Review by jkwatson — APPROVED** [review]
If it helps auto-instrumentation and doesn't hurt the end-user experience, I'm all in favor.

**Review by tylerbenson — APPROVED** [review]

File: api/src/main/java/io/opentelemetry/api/trace/SpanContext.java
Comment (tylerbenson) [inline api/src/main/java/io/opentelemetry/api/trace/SpanContext.java:17]: Does this still make sense on an interface?

**Comment by tylerbenson** [comment]
Can this please be included in the `0.10.0` release?

File: api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java:14]: Not sure why this class is needed, if you just move the "@AutoValue" annotation on the SpanContext you achieve the same thing, and just call directly the ctor of the AutoValue, am I missing something?

**Review by bogdandrutu — DISMISSED** [review]
Blocked until https://github.com/open-telemetry/opentelemetry-java/issues/1946

**Comment by jkwatson** [comment]
After further thinking, I'd prefer SpanContext stay not as an interface, but as a final class (or AutoValue'd class). It's a core piece of data for the API, and since it's data-only, having it be non-interface is preferable to me.

File: api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java
Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/api/trace/ImmutableSpanContext.java:14]: As far as I know, `AutoValue` doesn't support interfaces still.

**Comment by bogdandrutu** [comment]
@anuraaga @trask we are getting back to "auto-instrumentation" requires this, but not documented. Please make sure that this requirement is documented, not clear why this cannot happen, seems like a correct change.

**Review by jkwatson — APPROVED** [review]
Thanks!

**Review by trask — APPROVED** [review]

**Comment by trask** [comment]
@bogdandrutu can you unblock this PR, or comment on #1946 which you blocked this on?

**Comment by jkwatson** [comment]
> After further thinking, I'd prefer SpanContext stay not as an interface, but as a final class (or AutoValue'd class). It's a core piece of data for the API, and since it's data-only, having it be non-interface is preferable to me.

Note: although I usually favor data-as-classes, I'm on board with this change.

File: api/src/main/java/io/opentelemetry/api/trace/SpanContext.java
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/api/trace/SpanContext.java:116]: this will become more expensive after this change, we should calculate this only once since it is used couple of times per instance created.

File: api/src/main/java/io/opentelemetry/api/trace/SpanContext.java
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/api/trace/SpanContext.java:88]: Why removing the ability to memoized this call? It may get expensive. I would suggest to not default this and any method that was memoized before.

**Comment by bogdandrutu** [comment]
My concern still stays on a general level. If there was no auto-instrumentation project majority of the engineers have implemented this as a final class, so I am concerned that we do designed choices based on an external project need that may influence the overall usability and performance of the core product. This designed choices may affect us in the future in an unknown way.

I do understand the needs coming from auto-instrumentation, but you need to keep in mind that right now we need to be prepared to deal with cases where users will implement this in a "unexpected" way like having mutable objects inside the SpanContext, etc. this may cause problems for us, and we may have to deal with this.

File: api/src/main/java/io/opentelemetry/api/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/api/trace/SpanContext.java:88]: This is still memoized by the autovalue implementation. See ImmutableSpanContext above.

File: api/src/main/java/io/opentelemetry/api/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/api/trace/SpanContext.java:116]: Also, see above...this is still memoized by the implementation.

**Comment by anuraaga** [comment]
Thanks! Added a warning

**Comment by jkwatson** [comment]
@marcingrzejszczak this is potentially a breaking change

**Pull request merged**
